### ISSUE TITLE
[Proposal] Log rate limiting globally.

### ISSUE THREAD

**Issue by cyriltovena** [grafana/alloy#2065]
### Background

Right now the stage limit for log only apply locally to a single alloy instance. see https://grafana.com/docs/alloy/latest/reference/components/loki/loki.process/#stagelimit-block which is a bit impractical to use for rate limiting in large clusters where users want to apply rate limiting globally

### Proposal

Not sure if this is possible with cluster management but it would be nice to apply the rate limiting globally across all agents if they form a cluster.

We most likely want to use a different block such as `stage.global_limit`

**Comment by wildum** [comment]
One way to solve this with the current solution is to put the component in a [module](https://grafana.com/docs/alloy/latest/get-started/modules/). You can store the module in a git repo and all the Alloy instances will pull it regularly. Not only this allows you to change all the rate limits live across all the instances, but you can also change other parts of the pipeline this way.
Another similar approach is to use fleet management

**Comment by tg3rh** [comment]
@wildum would this also solve the issue of having a central rate limiting if alloy is deployed in clustering mode?

**Comment by wildum** [comment]
yes this works with clustering mode

**Comment by cyriltovena** [comment]
> One way to solve this with the current solution is to put the component in a [module](https://grafana.com/docs/alloy/latest/get-started/modules/). You can store the module in a git repo and all the Alloy instances will pull it regularly. Not only this allows you to change all the rate limits live across all the instances, but you can also change other parts of the pipeline this way. Another similar approach is to use fleet management

I don't think that does it. We need the rate limit counter to be global, this requires synchronization of the rate of data pushed to the same target.


cc @thampiotr

**Comment by thampiotr** [comment]
It wouldn't be a global rate limiting and we currently don't have capacity to store any additional data in the clustering library that would be shared across all instances. Clustering would also not be fit for such frequent updates that rate limiting requires.
If we were to solve it in Alloy it would likely be a large undertaking that would require some additional storage system, e.g. Redis. It may be easier to add rate limits in the downstream DB.

**Comment by wildum** [comment]
Hello, I'm moving this proposal to active. We discussed it in today's proposal meeting.
@cyriltovena what do you think about @thampiotr's suggestion to handle this in Loki?
Else another possible solution would be to have the cluster of Alloy sends all the logs to a central Alloy which would apply the rate limit before passing the logs downstream. But this only works if one Alloy can handle receiving all the logs and apply the rate limit.

**Comment by wildum** [comment]
Adding the proposal to likely closed for now because there is no easy way to solve this in Alloy and it's not a very requested feature. Happy to re-open it if there is a strong need for it

**Comment by thampiotr** [comment]
Moving this to likely decline because the current priority of this is "nice to have" but it would be significant effort and likely complicate the deployment further to make this a scalable feature.
